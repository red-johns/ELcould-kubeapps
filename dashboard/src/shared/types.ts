import { IAuthState } from "../reducers/auth";
import { IServiceCatalogState } from "../reducers/catalog";
import { IConfigState } from "../reducers/config";
import { INamespaceState } from "../reducers/namespace";
import { IAppRepositoryState } from "../reducers/repos";
import { hapi } from "./hapi/release";

// Allow defining multiple error classes
// tslint:disable:max-classes-per-file

class CustomError extends Error {
  // The constructor is defined so we can later on compare the returned object
  // via err.contructor  == FOO
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
export class ForbiddenError extends CustomError {}
export class UnauthorizedError extends CustomError {}

export class NotFoundError extends CustomError {}

export class ConflictError extends CustomError {}

export class UnprocessableEntity extends CustomError {}

export interface IRepo {
  name: string;
  url: string;
}

export interface IChartVersion {
  id: string;
  attributes: IChartVersionAttributes;
  relationships: {
    chart: {
      data: IChartAttributes;
    };
  };
}

export interface IChartVersionAttributes {
  version: string;
  app_version: string;
  created: string;
}

export interface IChart {
  id: string;
  attributes: IChartAttributes;
  relationships: {
    latestChartVersion: {
      data: IChartVersionAttributes;
    };
  };
}

export interface IChartAttributes {
  name: string;
  description: string;
  home?: string;
  icon?: string;
  keywords: string[];
  maintainers: Array<{
    name: string;
    email?: string;
  }>;
  repo: IRepo;
  sources: string[];
}

export interface IChartState {
  isFetching: boolean;
  selected: {
    error?: Error;
    version?: IChartVersion;
    versions: IChartVersion[];
    readme?: string;
    readmeError?: string;
    values?: string;
  };
  items: IChart[];
}

export interface IDeployment {
  metadata: {
    name: string;
    namespace: string;
  };
}

export interface IServiceSpec {
  ports: IPort[];
  clusterIP: string;
  type: string;
}

export interface IServiceStatus {
  loadBalancer: {
    ingress?: Array<{ ip?: string; hostname?: string }>;
  };
}

export interface IPort {
  name: string;
  port: number;
  protocol: string;
  targetPort: string;
  nodePort: string;
}

export interface IHTTPIngressPath {
  path: string;
}
export interface IIngressHTTP {
  paths: IHTTPIngressPath[];
}
export interface IIngressRule {
  host: string;
  http: IIngressHTTP;
}

export interface IIngressTLS {
  hosts: string[];
}

export interface IIngressSpec {
  rules: IIngressRule[];
  tls?: IIngressTLS[];
}

export interface IResource {
  apiVersion: string;
  kind: string;
  type: string;
  spec: any;
  status: any;
  metadata: {
    name: string;
    namespace: string;
    annotations: string;
    creationTimestamp: string;
    selfLink: string;
    resourceVersion: string;
    deletionTimestamp?: string;
    uid: string;
  };
}

export interface IOwnerReference {
  apiVersion: string;
  blockOwnerDeletion: boolean;
  kind: string;
  name: string;
  uid: string;
}

export interface ISecret {
  apiVersion: string;
  kind: string;
  type: string;
  data: { [s: string]: string };
  metadata: {
    name: string;
    namespace: string;
    annotations: string;
    creationTimestamp: string;
    selfLink: string;
    resourceVersion: string;
    deletionTimestamp?: string;
    uid: string;
  };
}

export interface IDeploymentStatus {
  replicas: number;
  updatedReplicas: number;
  availableReplicas: number;
}

export interface IAppState {
  isFetching: boolean;
  error?: Error;
  deleteError?: Error;
  // currently items are always Helm releases
  items: hapi.release.Release[];
  listingAll: boolean;
  listOverview?: IAppOverview[];
  selected?: hapi.release.Release;
}

export interface IStoreState {
  catalog: IServiceCatalogState;
  apps: IAppState;
  auth: IAuthState;
  charts: IChartState;
  config: IConfigState;
  kube: IKubeState;
  repos: IAppRepositoryState;
  deployment: IDeployment;
  namespace: INamespaceState;
}

interface IK8sResource {
  apiVersion: string;
  kind: string;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#objects */
export interface IK8sObject<M, SP, ST> extends IK8sResource {
  metadata: {
    annotations?: { [key: string]: string };
    creationTimestamp?: string;
    deletionTimestamp?: string | null;
    generation?: number;
    labels?: { [key: string]: string };
    name: string;
    namespace: string;
    resourceVersion?: string;
    uid: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
  spec?: SP;
  status?: ST;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#lists-and-simple-kinds */
export interface IK8sList<I, M> extends IK8sResource {
  items: I[];
  metadata?: {
    resourceVersion?: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
}

export interface IAppRepository
  extends IK8sObject<
      {
        clusterName: string;
        creationTimestamp: string;
        deletionGracePeriodSeconds: string | null;
        deletionTimestamp: string | null;
        resourceVersion: string;
        selfLink: string;
      },
      {
        type: string;
        url: string;
        auth: {
          header: {
            secretKeyRef: {
              name: string;
              key: string;
            };
          };
        };
      },
      undefined
    > {}

export interface IAppRepositoryList
  extends IK8sList<
      IAppRepository,
      {
        continue: string;
        resourceVersion: string;
        selfLink: string;
      }
    > {}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#response-status-kind */
export interface IStatus extends IK8sResource {
  kind: "Status";
  status: "Success" | "Failure";
  message: string;
  reason:
    | "BadRequest"
    | "Unauthorized"
    | "Forbidden"
    | "NotFound"
    | "AlreadyExists"
    | "Conflict"
    | "Invalid"
    | "Timeout"
    | "ServerTimeout"
    | "MethodNotAllowed"
    | "InternalError";
  details?: {
    kind?: string;
    name?: string;
    causes?: IStatusCause[] | string;
  };
}

interface IStatusCause {
  field: string;
  message: string;
  reason: string;
}

export interface IRouterPathname {
  router: {
    location: {
      pathname: string;
    };
  };
}

export interface IRuntimeVersion {
  name: string;
  version: string;
  runtimeImage: string;
  initImage: string;
}

export interface IRuntime {
  ID: string;
  versions: IRuntimeVersion[];
  depName: string;
  fileNameSuffix: string;
}

export interface IRBACRole {
  apiGroup: string;
  namespace?: string;
  clusterWide?: boolean;
  resource: string;
  verbs: string[];
}

export interface IAppOverview {
  releaseName: string;
  namespace: string;
  version: string;
  icon?: string;
  status: string;
  chart: string;
}

export interface IKubeItem<T> {
  isFetching: boolean;
  item?: T;
  error?: Error;
}

export interface IKubeState {
  items: { [s: string]: IKubeItem<IResource> };
}
