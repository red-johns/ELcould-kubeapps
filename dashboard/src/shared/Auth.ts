import Axios, { AxiosResponse } from "axios";
const AuthTokenKey = "kubeapps_auth_token";

export class Auth {
  public static getAuthToken() {
    return localStorage.getItem(AuthTokenKey);
  }

  public static setAuthToken(token: string) {
    return localStorage.setItem(AuthTokenKey, token);
  }

  public static unsetAuthToken() {
    return localStorage.removeItem(AuthTokenKey);
  }

  public static wsProtocols() {
    const token = this.getAuthToken();
    if (!token) {
      return [];
    }
    return [
      "base64url.bearer.authorization.k8s.io." + btoa(token).replace(/=*$/g, ""),
      "binary.k8s.io",
    ];
  }

  public static fetchOptions(): RequestInit {
    const headers = new Headers();
    headers.append("Authorization", `Bearer ${this.getAuthToken()}`);
    return {
      headers,
    };
  }

  // Throws an error if the token is invalid
  public static async validateToken(token: string) {
    try {
      await Axios.get("api/kube/", { headers: { Authorization: `Bearer ${token}` } });
    } catch (e) {
      const res = e.response as AxiosResponse;
      if (res.status === 401) {
        throw new Error("invalid token");
      }
    }
  }
}

export const axios = Axios.create();
