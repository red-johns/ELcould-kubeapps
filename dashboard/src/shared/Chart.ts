import axios from "axios";

export default class Chart {
  public static async getReadme(id: string, version: string) {
    const url = `${Chart.APIEndpoint}/assets/${id}/versions/${version}/README.md`;
    const { data } = await axios.get<string>(url);
    return data;
  }

  public static async getValues(id: string, version: string) {
    const url = `${Chart.APIEndpoint}/assets/${id}/versions/${version}/values.yaml`;
    const { data } = await axios.get<string>(url);
    return data;
  }

  public static async exists(id: string, version: string, repo: string) {
    const url = `${Chart.APIEndpoint}/charts/${repo}/${id}/versions/${version}`;
    try {
      await axios.get<string>(url);
    } catch (e) {
      return false;
    }
    return true;
  }

  private static APIEndpoint: string = "api/chartsvc/v1";
}
