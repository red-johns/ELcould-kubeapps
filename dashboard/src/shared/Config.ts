import axios from "axios";

// IConfig is the configuration for Kubeapps
export interface IConfig {
  namespace: string;
  appVersion: string;
}

export default class Config {
  public static async getConfig() {
    const url = Config.APIEndpoint;
    const { data } = await axios.get<IConfig>(url);

    // Development environment config overrides
    // TODO(miguel) Rename env variable to KUBEAPPS_NAMESPACE once/if we eject create-react-app
    // Currently we are using REACT_APP_* because it's the only way to inject env variables in a sealed setup.
    // Please note that this env variable gets mapped in the run command in the package.json file
    if (process.env.NODE_ENV !== "production" && process.env.REACT_APP_KUBEAPPS_NS) {
      data.namespace = process.env.REACT_APP_KUBEAPPS_NS;
      data.appVersion = "DEVEL";
    }

    return data;
  }

  private static APIEndpoint: string = "config.json";
}
