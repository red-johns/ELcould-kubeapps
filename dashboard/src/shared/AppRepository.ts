import { axios } from "./Auth";
import { IAppRepository, IAppRepositoryList } from "./types";

export class AppRepository {
  public static async list(namespace: string) {
    const { data } = await axios.get<IAppRepositoryList>(AppRepository.getResourceLink(namespace));
    return data;
  }

  public static async get(name: string, namespace: string) {
    const { data } = await axios.get(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async update(name: string, namespace: string, newApp: IAppRepository) {
    const { data } = await axios.put(AppRepository.getSelfLink(name, namespace), newApp);
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const { data } = await axios.delete(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async create(name: string, namespace: string, url: string, auth: any) {
    const { data } = await axios.post<IAppRepository>(AppRepository.getResourceLink(namespace), {
      apiVersion: "kubeapps.com/v1alpha1",
      kind: "AppRepository",
      metadata: {
        name,
      },
      spec: { auth, type: "helm", url },
    });
    return data;
  }

  private static APIBase: string = "api/kube";
  private static APIEndpoint: string = `${AppRepository.APIBase}/apis/kubeapps.com/v1alpha1`;
  private static getResourceLink(namespace?: string): string {
    return `${AppRepository.APIEndpoint}/${
      namespace ? `namespaces/${namespace}/` : ""
    }apprepositories`;
  }
  private static getSelfLink(name: string, namespace: string): string {
    return `${AppRepository.APIEndpoint}/namespaces/${namespace}/apprepositories/${name}`;
  }
}
