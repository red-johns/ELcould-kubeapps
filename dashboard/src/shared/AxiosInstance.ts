import { AxiosError, AxiosInstance, AxiosRequestConfig } from "axios";
import { Action, Store } from "redux";
import { ThunkDispatch } from "redux-thunk";
import actions from "../actions";
import { Auth } from "./Auth";
import {
  ConflictError,
  ForbiddenError,
  IStoreState,
  NotFoundError,
  UnauthorizedError,
  UnprocessableEntity,
} from "./types";

// createAxiosInterceptors will configure a set of interceptors to a provided axios instance,
// relying also on an external redux store for action dispatching
export function createAxiosInterceptors(axios: AxiosInstance, store: Store<IStoreState>) {
  axios.interceptors.request.use((config: AxiosRequestConfig) => {
    const authToken = Auth.getAuthToken();
    if (authToken) {
      config.headers.Authorization = `Bearer ${authToken}`;
    }
    return config;
  });
  axios.interceptors.response.use(
    response => response,
    e => {
      const err: AxiosError = e;
      let message = err.message;
      if (err.response && err.response.data.message) {
        message = err.response.data.message;
      }
      switch (err.response && err.response.status) {
        case 401:
          // Global action dispatch to log the user out
          const dispatch = store.dispatch as ThunkDispatch<IStoreState, null, Action>;
          dispatch(actions.auth.authenticationError(message));
          dispatch(actions.auth.logout());
          return Promise.reject(new UnauthorizedError(message));
        case 403:
          return Promise.reject(new ForbiddenError(message));
        case 404:
          return Promise.reject(new NotFoundError(message));
        case 409:
          return Promise.reject(new ConflictError(message));
        case 422:
          return Promise.reject(new UnprocessableEntity(message));
        default:
          return Promise.reject(new Error(message));
      }
    },
  );
}
