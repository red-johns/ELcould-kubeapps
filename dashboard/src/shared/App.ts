import { AppRepository } from "./AppRepository";
import { axios } from "./Auth";
import { hapi } from "./hapi/release";
import { IAppOverview, IChartVersion } from "./types";

export const TILLER_PROXY_ROOT_URL = "api/tiller-deploy/v1";

export class App {
  public static getResourceURL(namespace?: string, name?: string, query?: string) {
    let url = TILLER_PROXY_ROOT_URL;
    if (namespace) {
      url += `/namespaces/${namespace}`;
    }
    url += "/releases";
    if (name) {
      url += `/${name}`;
    }
    if (query) {
      url += `?${query}`;
    }
    return url;
  }

  public static async create(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const repo = await AppRepository.get(chartAttrs.repo.name, kubeappsNamespace);
    const auth = repo.spec.auth;
    const endpoint = App.getResourceURL(namespace);
    const { data } = await axios.post(endpoint, {
      auth,
      chartName: chartAttrs.name,
      releaseName,
      repoUrl: chartAttrs.repo.url,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const repo = await AppRepository.get(chartAttrs.repo.name, kubeappsNamespace);
    const auth = repo.spec.auth;
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axios.put(endpoint, {
      auth,
      chartName: chartAttrs.name,
      releaseName,
      repoUrl: chartAttrs.repo.url,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async delete(releaseName: string, namespace: string, purge: boolean) {
    let purgeQuery;
    if (purge) {
      purgeQuery = "purge=true";
    }
    const { data } = await axios.delete(App.getResourceURL(namespace, releaseName, purgeQuery));
    return data;
  }

  public static async listApps(namespace?: string, allStatuses?: boolean) {
    let query;
    if (allStatuses) {
      query = "statuses=all";
    }
    const { data } = await axios.get<{ data: IAppOverview[] }>(
      App.getResourceURL(namespace, undefined, query),
    );
    return data.data;
  }

  public static async getRelease(namespace: string, name: string) {
    const { data } = await axios.get<{ data: hapi.release.Release }>(
      this.getResourceURL(namespace, name),
    );
    return data.data;
  }
}
