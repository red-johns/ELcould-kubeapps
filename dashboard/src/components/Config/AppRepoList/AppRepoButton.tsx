import * as React from "react";
import * as Modal from "react-modal";
import { Redirect } from "react-router";

import { IRBACRole } from "../../../shared/types";
import ErrorSelector from "../../ErrorAlert/ErrorSelector";
import "./AppRepo.css";

interface IAppRepoFormProps {
  name: string;
  url: string;
  authHeader: string;
  customCA: string;
  message?: string;
  redirectTo?: string;
  install: (name: string, url: string, authHeader: string, customCA: string) => Promise<boolean>;
  update: (values: { name?: string; url?: string; authHeader?: string; customCA?: string }) => void;
  onAfterInstall?: () => Promise<any>;
}

const RequiredRBACRoles: IRBACRole[] = [
  {
    apiGroup: "kubeapps.com",
    resource: "apprepositories",
    verbs: ["create"],
  },
  {
    apiGroup: "",
    resource: "secrets",
    verbs: ["create"],
  },
];

export const AppRepoForm = (props: IAppRepoFormProps) => {
  const { name, url, authHeader, customCA, update, install, onAfterInstall } = props;
  const handleInstallClick = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const installed = await install(name, url, authHeader, customCA);
    if (installed && onAfterInstall) {
      await onAfterInstall();
    }
  };
  const handleNameChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    update({ name: e.target.value });
  const handleURLChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    update({ url: e.target.value });
  const handleAuthHeaderChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    update({ authHeader: e.target.value });
  const handleCustomCAChange = (e: React.ChangeEvent<HTMLTextAreaElement>) =>
    update({ customCA: e.target.value });
  return (
    <form className="container padding-b-bigger" onSubmit={handleInstallClick}>
      <div className="row">
        <div className="col-12">
          <div>
            <h2>Add an App Repository</h2>
          </div>
          <div>
            <label>
              <span>Name:</span>
              <input
                type="text"
                placeholder="example"
                value={name}
                onChange={handleNameChange}
                required={true}
                pattern="[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*"
                title="Use lower case alphanumeric characters, '-' or '.'"
              />
            </label>
          </div>
          <div>
            <label>
              <span>URL:</span>
              <input
                type="url"
                placeholder="https://charts.example.com/stable"
                value={url}
                onChange={handleURLChange}
                required={true}
              />
            </label>
          </div>
          <div>
            <label>
              <span>Authorization Header (optional):</span>
              <input
                type="text"
                placeholder="Bearer xrxNcWghpRLdcPHFgVRM73rr4N7qjvjm"
                value={authHeader}
                onChange={handleAuthHeaderChange}
              />
            </label>
          </div>
          <div>
            <label>
              <span>Custom CA Certificate (optional):</span>
              <pre className="CertContainer">
                <textarea
                  className="CertContent"
                  rows={4}
                  placeholder={
                    "-----BEGIN CERTIFICATE-----\n" + "...\n" + "-----END CERTIFICATE-----"
                  }
                  value={customCA}
                  onChange={handleCustomCAChange}
                />
              </pre>
            </label>
          </div>
          <div>
            <button className="button button-primary" type="submit">
              Install Repo
            </button>
          </div>
          {props.redirectTo && <Redirect to={props.redirectTo} />}
        </div>
      </div>
    </form>
  );
};

interface IAppRepoAddButtonProps {
  error?: Error;
  install: (name: string, url: string, authHeader: string, customCA: string) => Promise<boolean>;
  redirectTo?: string;
  kubeappsNamespace: string;
}
interface IAppRepoAddButtonState {
  authHeader: string;
  customCA: string;
  lastSubmittedName: string;
  modalIsOpen: boolean;
  name: string;
  url: string;
}

export class AppRepoAddButton extends React.Component<
  IAppRepoAddButtonProps,
  IAppRepoAddButtonState
> {
  public state = {
    authHeader: "",
    customCA: "",
    error: undefined,
    lastSubmittedName: "",
    modalIsOpen: false,
    name: "",
    url: "",
  };

  public render() {
    const { redirectTo } = this.props;
    const { name, url, authHeader, customCA } = this.state;
    return (
      <div className="AppRepoAddButton">
        <button className="button button-primary" onClick={this.openModal}>
          Add App Repository
        </button>
        <Modal
          isOpen={this.state.modalIsOpen}
          onRequestClose={this.closeModal}
          contentLabel="Modal"
        >
          {this.props.error && (
            <ErrorSelector
              error={this.props.error}
              defaultRequiredRBACRoles={{ create: RequiredRBACRoles }}
              action="create"
              namespace={this.props.kubeappsNamespace}
              resource={`App Repository ${this.state.lastSubmittedName}`}
            />
          )}
          <AppRepoForm
            name={name}
            url={url}
            authHeader={authHeader}
            customCA={customCA}
            update={this.updateValues}
            install={this.install}
            onAfterInstall={this.closeModal}
          />
        </Modal>
        {redirectTo && <Redirect to={redirectTo} />}
      </div>
    );
  }

  private closeModal = async () => this.setState({ modalIsOpen: false });
  private install = (name: string, url: string, authHeader: string, customCA: string) => {
    // Store last submitted name to show it in an error if needed
    this.setState({ lastSubmittedName: this.state.name });
    return this.props.install(name, url, authHeader, customCA);
  };
  private openModal = async () => this.setState({ modalIsOpen: true });
  private updateValues = async (values: {
    name: string;
    url: string;
    authHeader: string;
    customCA: string;
  }) => this.setState({ ...values });
}
