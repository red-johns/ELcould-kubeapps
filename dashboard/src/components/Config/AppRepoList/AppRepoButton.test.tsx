import { mount } from "enzyme";
import * as React from "react";
import * as ReactModal from "react-modal";
import { ConflictError, UnprocessableEntity } from "../../../shared/types";
import ErrorSelector from "../../ErrorAlert/ErrorSelector";
import { AppRepoAddButton, AppRepoForm } from "./AppRepoButton";

const defaultProps = {
  install: jest.fn(),
  kubeappsNamespace: "kubeapps",
};

it("should open a modal with the repository form", () => {
  const wrapper = mount(<AppRepoAddButton {...defaultProps} />);
  ReactModal.setAppElement(document.createElement("div"));
  wrapper.setState({ modalIsOpen: true });
  expect(wrapper).toMatchSnapshot();
});

it("should install a repository", done => {
  const install = jest.fn(() => true);
  const wrapper = mount(<AppRepoAddButton {...defaultProps} install={install} />);
  ReactModal.setAppElement(document.createElement("div"));
  wrapper.setState({
    modalIsOpen: true,
    name: "my-repo",
    url: "http://foo.bar",
    authHeader: "foo",
    customCA: "bar",
  });
  wrapper.update();

  const button = wrapper.find(AppRepoForm).find(".button");
  button.simulate("submit");

  expect(install).toBeCalledWith("my-repo", "http://foo.bar", "foo", "bar");
  // Wait for the Modal to be closed
  setTimeout(() => {
    expect(wrapper.state("modalIsOpen")).toBe(false);
    done();
  }, 1);
});

describe("render error", () => {
  it("renders a conflict error", () => {
    const wrapper = mount(<AppRepoAddButton {...defaultProps} />);
    ReactModal.setAppElement(document.createElement("div"));
    wrapper.setState({ modalIsOpen: true, name: "my-repo" });
    wrapper.update();

    const button = wrapper.find(AppRepoForm).find(".button");
    button.simulate("submit");
    wrapper.setProps({ error: new ConflictError("already exists!") });

    expect(wrapper.find(ErrorSelector).text()).toContain(
      "App Repository my-repo already exists, try a different name.",
    );
    // Now changing the name should not change the error message
    wrapper.setState({ name: "my-app-2" });
    wrapper.update();
    expect(wrapper.find(ErrorSelector).text()).toContain(
      "App Repository my-repo already exists, try a different name.",
    );
  });

  it("renders an 'unprocessable entity' error", () => {
    const wrapper = mount(<AppRepoAddButton {...defaultProps} />);
    ReactModal.setAppElement(document.createElement("div"));
    wrapper.setState({ modalIsOpen: true, name: "my-repo" });
    wrapper.update();

    const button = wrapper.find(AppRepoForm).find(".button");
    button.simulate("submit");
    wrapper.setProps({ error: new UnprocessableEntity("cannot process this!") });

    expect(wrapper.find(ErrorSelector).text()).toContain(
      "Something went wrong processing App Repository my-repo",
    );
    expect(wrapper.find(ErrorSelector).text()).toContain("cannot process this!");
  });
});
