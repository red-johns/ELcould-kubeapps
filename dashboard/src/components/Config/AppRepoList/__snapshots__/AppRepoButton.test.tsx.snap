// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should open a modal with the repository form 1`] = `
<AppRepoAddButton
  install={[MockFunction]}
  kubeappsNamespace="kubeapps"
>
  <div
    className="AppRepoAddButton"
  >
    <button
      className="button button-primary"
      onClick={[Function]}
    >
      Add App Repository
    </button>
    <Modal
      ariaHideApp={true}
      bodyOpenClassName="ReactModal__Body--open"
      closeTimeoutMS={0}
      contentLabel="Modal"
      isOpen={true}
      onRequestClose={[Function]}
      parentSelector={[Function]}
      portalClassName="ReactModalPortal"
      shouldCloseOnEsc={true}
      shouldCloseOnOverlayClick={true}
      shouldFocusAfterRender={true}
      shouldReturnFocusAfterClose={true}
    >
      <Portal
        containerInfo={
          <div
            class="ReactModalPortal"
          >
            <div
              aria-modal="true"
              class="ReactModal__Overlay ReactModal__Overlay--after-open"
              style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
            >
              <div
                aria-label="Modal"
                class="ReactModal__Content ReactModal__Content--after-open"
                style="position: absolute; top: 40px; left: 40px; right: 40px; bottom: 40px; border: 1px solid #ccc; background: rgb(255, 255, 255); overflow: auto; border-radius: 4px; outline: none; padding: 20px;"
                tabindex="-1"
              >
                <form
                  class="container padding-b-bigger"
                >
                  <div
                    class="row"
                  >
                    <div
                      class="col-12"
                    >
                      <div>
                        <h2>
                          Add an App Repository
                        </h2>
                      </div>
                      <div>
                        <label>
                          <span>
                            Name:
                          </span>
                          <input
                            pattern="[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*"
                            placeholder="example"
                            required=""
                            title="Use lower case alphanumeric characters, '-' or '.'"
                            type="text"
                            value=""
                          />
                        </label>
                      </div>
                      <div>
                        <label>
                          <span>
                            URL:
                          </span>
                          <input
                            placeholder="https://charts.example.com/stable"
                            required=""
                            type="url"
                            value=""
                          />
                        </label>
                      </div>
                      <div>
                        <label>
                          <span>
                            Authorization Header (optional):
                          </span>
                          <input
                            placeholder="Bearer xrxNcWghpRLdcPHFgVRM73rr4N7qjvjm"
                            type="text"
                            value=""
                          />
                        </label>
                      </div>
                      <div>
                        <label>
                          <span>
                            Custom CA Certificate (optional):
                          </span>
                          <pre
                            class="CertContainer"
                          >
                            <textarea
                              class="CertContent"
                              placeholder="-----BEGIN CERTIFICATE-----
...
-----END CERTIFICATE-----"
                              rows="4"
                            />
                          </pre>
                        </label>
                      </div>
                      <div>
                        <button
                          class="button button-primary"
                          type="submit"
                        >
                          Install Repo
                        </button>
                      </div>
                    </div>
                  </div>
                </form>
              </div>
            </div>
          </div>
        }
      >
        <ModalPortal
          ariaHideApp={true}
          bodyOpenClassName="ReactModal__Body--open"
          closeTimeoutMS={0}
          contentLabel="Modal"
          defaultStyles={
            Object {
              "content": Object {
                "WebkitOverflowScrolling": "touch",
                "background": "#fff",
                "border": "1px solid #ccc",
                "borderRadius": "4px",
                "bottom": "40px",
                "left": "40px",
                "outline": "none",
                "overflow": "auto",
                "padding": "20px",
                "position": "absolute",
                "right": "40px",
                "top": "40px",
              },
              "overlay": Object {
                "backgroundColor": "rgba(255, 255, 255, 0.75)",
                "bottom": 0,
                "left": 0,
                "position": "fixed",
                "right": 0,
                "top": 0,
              },
            }
          }
          isOpen={true}
          onRequestClose={[Function]}
          parentSelector={[Function]}
          portalClassName="ReactModalPortal"
          shouldCloseOnEsc={true}
          shouldCloseOnOverlayClick={true}
          shouldFocusAfterRender={true}
          shouldReturnFocusAfterClose={true}
          style={
            Object {
              "content": Object {},
              "overlay": Object {},
            }
          }
        />
      </Portal>
    </Modal>
  </div>
</AppRepoAddButton>
`;
